File: leetcode/199.binary-tree-right-side-view.cpp
//  Tag: Tree, Depth-First Search, Breadth-First Search, Binary Tree
//  Time: O(N)
//  Space: O(W)
//  Ref: -
//  Note: -

//  Given the root of a binary tree, imagine yourself standing on the right side of it, return the values of the nodes you can see ordered from top to bottom.
//   
//  Example 1:
//  
//  
//  Input: root = [1,2,3,null,5,null,4]
//  Output: [1,3,4]
//  
//  Example 2:
//  
//  Input: root = [1,null,3]
//  Output: [1,3]
//  
//  Example 3:
//  
//  Input: root = []
//  Output: []
//  
//   
//  Constraints:
//  
//  The number of nodes in the tree is in the range [0, 100].
//  -100 <= Node.val <= 100
//  
//  

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        if (!root) {
            return vector<int>{};
        }

        vector<int> res;
        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty()) {
            int size = q.size();
            for (int i = 0; i < size; i++) {
                TreeNode* cur = q.front();
                q.pop();
                if (cur->left) {
                    q.push(cur->left);
                }
                if (cur->right) {
                    q.push(cur->right);
                }
                if (i == size - 1) {
                    res.push_back(cur->val);
                }
            }
        }
        return res;
    }
};

class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> res;
        helper(root, 1, res);
        return res;
    }

    void helper(TreeNode *cur, int level, vector<int> &res) {
        if (!cur) {
            return;
        }

        if (res.size() < level) {
            res.push_back(cur->val);
        }

        helper(cur->right, level + 1, res);
        helper(cur->left, level + 1, res);
    }
};

File: leetcode/199.binary-tree-right-side-view.py
#  Tag: Tree, Depth-First Search, Breadth-First Search, Binary Tree
#  Time: O(N)
#  Space: O(W)
#  Ref: -
#  Note: -

#  Given the root of a binary tree, imagine yourself standing on the right side of it, return the values of the nodes you can see ordered from top to bottom.
#   
#  Example 1:
#  
#  
#  Input: root = [1,2,3,null,5,null,4]
#  Output: [1,3,4]
#  
#  Example 2:
#  
#  Input: root = [1,null,3]
#  Output: [1,3]
#  
#  Example 3:
#  
#  Input: root = []
#  Output: []
#  
#   
#  Constraints:
#  
#  The number of nodes in the tree is in the range [0, 100].
#  -100 <= Node.val <= 100
#  
#  

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if root is None:
            return []

        res = []
        q = deque([root])
        while len(q) > 0:
            size = len(q)
            for i in range(size):
                cur = q.popleft()
                if cur.left is not None:
                    q.append(cur.left)
                if cur.right is not None:
                    q.append(cur.right)

                if i == size - 1:
                    res.append(cur.val)

        return res
    
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        self.helper(root, 1, res)
        return res

    def helper(self, node: TreeNode, level: int, res:list):
        if node is None:
            return

        if len(res) < level:
            res.append(node.val)

        self.helper(node.right, level + 1, res)
        self.helper(node.left, level + 1, res)